### PR TITLE
[develop] Performance: Add preloading config adapter

### LINKED ISSUE #4520: Performance: Load all the config at once

### PR DESCRIPTION
**Pull request by MrPetovan**
Closes #4520

In the same spirit as the cache drivers, I added a new config adapter for people wanting to get extreme performances with Friendica.

With the new preload Config adapter, the best gains will show in the `settings` page but also in the `network` loads.

### CODE REVIEW

File: src/Core/Config/PreloadConfigAdapter.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+				$a->config[$cat][$k] = $value;
+			}
+		}
+		dba::close($configs);
+
+		$this->config_loaded = true;
+	}
+
+	public function get($cat, $k, $default_value = null, $refresh = false)
+	{
```
Comment (annando) [inline src/Core/Config/PreloadConfigAdapter.php]: ```$refresh = true``` has to be coded as well. There are places in the code that depend on always fetching the latest config value.

For example "src/Util/ExAuth.php" is a long running process - it is started with an ejabberd and then runs for several hours/days/weeks.

Same goes with "scripts/worker.php". It has to check whether the system is in maintenance mode. Then it has to stop immediately. (BTW: There is currently a bug in the worker)

File: src/Core/Config/PreloadPConfigAdapter.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+			$a->config[$uid][$cat][$k] = $value;
+		}
+		dba::close($pconfigs);
+
+		$this->config_loaded = true;
+	}
+
+	public function get($uid, $cat, $k, $default_value = null, $refresh = false)
+	{
```
Comment (annando) [inline src/Core/Config/PreloadPConfigAdapter.php]: See my comment above. We have to be able to fetch completely new content.

File: src/Core/Config/PreloadConfigAdapter.php (hunk 1 above)
Comment (MrPetovan) [reply src/Core/Config/PreloadConfigAdapter.php]: I see, thanks for your vigilance.

Comment (MrPetovan) [reply src/Core/Config/PreloadConfigAdapter.php]: Done.

File: src/Core/Config/PreloadPConfigAdapter.php (hunk 2 above)
Comment (MrPetovan) [reply src/Core/Config/PreloadPConfigAdapter.php]: Done.

**Comment by annando** [comment]
It seems as if the main difference between the two classes is the preloading. The other parts look very similar. For one of the classes you created functions like ``deletePreloadedValue``` which could be helpful for the other class as well.

I would prefer having these class calls in the main class, to avoid code duplication.

**Comment by MrPetovan** [comment]
Alright, this branch is now checked out on my server as of this morning. This means I can't check #4543 at the same time.

**Comment by MrPetovan** [comment]
Those calls should even be in the `App` class because it's interacting with a class property, but I didn't want to change too much. I'll do that tonight.

**Comment by annando** [comment]
Why not the "Config" class?

**Comment by MrPetovan** [comment]
Because it isn’t a real class, just a namespaced list of static methods that modify `App` class members.

From there it makes more sense to let the `App` class modify its own members.

**Comment by annando** [comment]
Okay. I don't understand this concept, but I trust you.

**Comment by MrPetovan** [comment]
Let's say `App` is a restaurant, and `Config` is a food truck. Currently, `Config` is selling food it takes from the `App` kitchen through the window. Why not sell the food directly from `App` since it's made in its kitchen? We don't need such a middleman.

**Comment by MrPetovan** [comment]
I started moving calls to the `App` class but it doesn't look like I will be able to use them much in the current Config adapter, it uses `$s->config` in very specific ways with the '!<unset>!` values.

**Comment by MrPetovan** [comment]
> I would prefer having these class calls in the main class, to avoid code duplication.

I did my best without touching to the original Config Adapter implementation I'm not sure to fully grasp. Ideally, at some point in the future I'd like to set the config preloader as the default adapter, but for that we have to get rid of all the update_xxx config variables. Since the update process is linear and one-way, it doesn't make sense to keep one flag for each update number.

**Comment by MrPetovan** [comment]
I'm currently tracking an issue where the Config array isn't preloaded during a worker execution. Please don't merge yet.

**Comment by annando** [comment]
When you are checking this whole config stuff, please consider testing the scripts (like dbstructure.php) as well.

**Comment by MrPetovan** [comment]
Alright, I found the issue, there were more application entry points than I thought. This is working properly now.

**Pull request merged**